### ISSUE TITLE
SSR / Simple firestore get() query not rendered

### ISSUE THREAD

**Issue by vroussel35** [firebase/firebase-js-sdk#3616]
### [REQUIRED] Describe your environment

  * Operating System version: Mac OS / SSR context
  * Browser version: N/A
  * Firebase SDK version:  7.15.2 and above
  * Firebase Product: Firestore

### [REQUIRED] Describe the problem

When making a simple firestore query on the server side (Angular 9+ / SSR), the query takes up to 5 minutes to "complete".

As an example, a get() call is executed (the console.log is well shown on the console) but after, the execution is kept hang for several minutes, before the page being rendered. 

```
firebase.firestore().collection('homepages_public').doc('32a7hGeVfgwJ0v2eEBw3').get().then((result) => {
 console.log(result.id);
});
```

I suspect there is an ongoing connection to Firestore which is persistent, which makes the Angular Zone waiting for this connection to be completed / timed out, hence rendering only after few minutes.

#### Steps to reproduce:

- Download this project : https://github.com/vroussel35/error-ssr-firestore-query
- Execute `npm install`
- Execute `npm run dev:ssr`
- Browse [http://localhost:4200](http://localhost:4200) and look at the server console.

**Comment by wu-hui** [comment]
Hey, thanks for providing reproduction steps!

I gave your repo a try, and I can confirm by turning on debug logging (`firebase.setLogLevel('debug');` when you initializeApp), that the connection to firestore is fine. The query went through and the result is returned. The connection is closed about 1 minute later, and then server is trying to create `FirebaseService` again, which leads to a duplicate App error.

I am not familiar with Angular and SSR, but I suspect something from that environment is trying to manage the firebase services. Maybe you should consider making `FirebaseService` a singleton somehow?

Let me know if this makes sense.

**Comment by vroussel35** [comment]
@wu-hui Thanks for your answer.

The timeouts you have seen are not linked to this issue: It's the default (timeout after 2 min) value of the underlying objects used by the `@nguniversal/builders:ssr-dev-server` builder when running the `npm run dev:ssr` command.

The timeout is happening because "something" (firestore?) is hanging on the server side and not letting the server completing/sending its answer to the client browser.

This timeout value cannot be configured so I've **updated the repository** and **added below the new reproduction steps** to focus only on the issue.

**Steps to reproduce:**
1. Download this project : https://github.com/vroussel35/error-ssr-firestore-query
2. Execute npm install
3. npm run build:ssr
4. npm run serve:ssr
5. Browse http://localhost:4000 and look at the server console.

If we look at what is happening here:

1. The Node/Express server is receiving the browser request
2. The service `FirebaseService` (it's already a singleton as per the Angular framework) calls the initializeApp() function.
3. The firestore query is made and the result is received within ~200/300ms
3a. After 1 min, we can see that a GRPC stream is closed
```
[2020-08-14T15:50:53.202Z]  @firebase/firestore: Firestore (7.18.0): Connection GRPC stream closed locally via close().
[2020-08-14T15:50:53.203Z]  @firebase/firestore: Firestore (7.18.0): PersistentStream stream callback skipped by getCloseGuardedDispatcher.
[2020-08-14T15:50:53.237Z]  @firebase/firestore: Firestore (7.18.0): Connection GRPC stream ended.
```
but still.... the execution on the server is still hanging... (and the client browser is still waiting for its answer...)

4. Then... after 5 min, the answer is sent back to the client browser - I suspect here it's the default timeout Chrome browser settings which is triggered here (I'm testing using Chrome).

If we remove the call to firestore, the page is well rendered in few ms. If we make this call to firestore, the page is not rendered.

**Comment by wu-hui** [comment]
Thanks..after trying the repo, i think Firestore is doing properly, it looks like this is an issue from angular SSR instead.

See: https://github.com/angular/angularfire/issues/2420

Although you are not using angularfire, it looks like the same issue, and this seems to be caused by `rxjs` running in SSR. 

Sorry, I am afraid I cannot provide a solution for you. Hopefully you will find a walkaround there.

**Comment by vroussel35** [comment]
@wu-hui : thanks for your input.

I don't agree with your analysis => the issue you have mentioned is related to `AngularFirestore` and `AngularFirestore` is working with SSR. What is true is there are some cases where `rxjs` causes some problems with `AngularFirestore` but a simple call using the `AngularFirestore` package works when running in SSR. This is a separate issue.

@jamesdaniels Because you seems to have solved the issue in the `AngularFirestore` package, are you able to add some inputs that can give @wu-hui some hints on where the issue/root cause could be?

**Comment by jamesdaniels** [comment]
The JS SDK has a lot of timers and other side effects, persistent connections, etc. that are not Zone-patched. That's the root cause here.

I think it's out of scope to expect framework-agnostic libraries to Zone-js patch themselves.

**Comment by vroussel35** [comment]
Thanks @jamesdaniels.

Does it mean there is no solution - as of today - for using firestore in SSR mode using Promises?

We can still use the AngularFirestore (a little bit buggy with issues when using compound queries (see this issue https://github.com/angular/angularfire/issues/2420) which is Observable based but if we want to use Promises, it seems there is no solution available.

Don't get me wrong - I just want to find a solution for our project where SSR is a key business factor to have and since more than 6 months, we are stuck with Angular 8 as the solutions are buggy or not available.

**Comment by sagrawal31** [comment]
I'm stuck on this like hell. Had to change every code to use observable and wrap everything as a promise.

**Comment by Feiyang1** [comment]
@vroussel35 We are working on a new version of the Firebase SDK(including `Firestore`) where we will provide a new entry point in `Firestore` that provides light weight promise based APIs that uses the REST endpoints directly. see https://github.com/firebase/firebase-js-sdk/issues/2241#issuecomment-683130414 

It should help you solve the SSR use case once it is released, though no promise on when it will happen. Sorry!

**Comment by DellaBitta** [comment]
It's been a while. Checking in to see if this still is still an issue for folks. Thanks!
